### ISSUE TITLE
Temperature flux statistics in BdyLayerStatistics.C

### ISSUE THREAD

**Issue by lawrenceccheung** [Exawind/nalu-wind#499]
## Detailed description of the bug 

I'm getting some strange results for the temperature flux quantities (`temperature_resolved_flux`) coming out of the ABL statistics routines:
![image](https://user-images.githubusercontent.com/15526007/73031325-a3eec100-3df0-11ea-907f-5741c7532530.png)

For a neutral ABL, I would expect that the <T'w'> flux to be basically zero, and the magnitudes of these values are much too high for fluctuating quantities.   However, the temperature T(z) profiles seem to be computed fine.

Alan Hsieh had earlier reported reasonable statistics for the temperature flux, so I'm wondering if there were any changes recently.

Going through `impl_compute_temperature_stats()` in [BdyLayerStatistics.C](https://github.com/Exawind/nalu-wind/blob/master/src/wind_energy/BdyLayerStatistics.C#L541),
```c++
      const int offset = ih * ndim;
      for (int d=0; d < ndim; ++d) {
        Kokkos::atomic_add(
          &d_thetaSFSBarAvg(offset + d), (thetaSFS.get(mi, 0) * dVol));
        Kokkos::atomic_add(
          &d_thetaUjBarAvg(offset + d), (thetaUj.get(mi, 0) * dVol));
        Kokkos::atomic_add(
          &d_thetaUjAvg(offset + d),
          (rho * theta.get(mi, 0) * velocity.get(mi, 0) * dVol));
      }
```
it looks like the `Kokkos::atomic_add` only happens over the x direction of the quantities.  Should it be over the *d* direction in the second argument, like this:

```c++
        Kokkos::atomic_add( &d_thetaSFSBarAvg(offset + d), (thetaSFS.get(mi, d) * dVol));
        Kokkos::atomic_add( &d_thetaUjBarAvg(offset + d), (thetaUj.get(mi, d) * dVol));
        Kokkos::atomic_add( &d_thetaUjAvg(offset + d), (rho * theta.get(mi, 0) * velocity.get(mi, d) * dVol));
```

Not sure if that is the relevant issue, or if there are other things that would lead to this.  Also I'm not familiar with Kokkos programming so I could use some help in debugging.

### Details to reproduce bug/build error

1. SHA ID of the Nalu-Wind commit you are using 
1. Configure command [attach file]
1. CMake configure output [attach file]
1. Output of make step [attach file]
1. Details of your simulation (input file, mesh description, etc.)
1. Stack trace of the runtime error output

**Comment by sayerhs** [comment]
@lawrenceccheung yes. You are correct that the second argument to the`ngp::Field::get` method should be the index `d`. Please submit a bug fix.

**Comment by lawrenceccheung** [comment]
@sayerhs Will do.  Let me try this change, and if it solves the problem, I'll put in a bug fix.

**Comment by lucbv** [comment]
@lawrenceccheung if you can add a test that fails with the current bug that would be extra nice : )

**Comment by lawrenceccheung** [comment]
@sayerhs Okay, so changing the second argument to `ngp::Field::get` solved my immediate problem, and now `temperature_resolved_flux` looks much more reasonable.  

However, there's another issue in the ABL stats which has come up.  It looks like the `'velocity_tavg'` profiles are off from the `'velocity'` by a factor of rho (for simulations where rho != 1.0).  I think it's also throwing off the profiles in `temperature_resolved_flux_tavg`.

Alan Hsieh (@alhs6577) mentioned that this discrepancy doesn't appear in older versions of Nalu-Wind (from earlier in 2019).

I'm looking into where this factor could come from -- maybe from the `velBarAvg` computation in `BdyLayerStatistics.C` or from `compute_averages()` in `TurbulenceAveragingPostProcessing.C`.  Not sure yet which one it is yet.

**Comment by lawrenceccheung** [comment]
Alan (@alhs6577) and I also think that it would be a good idea to do a full comparison of the ABL statistics between the older Nalu-Wind version and one from the post-NGP modifications.

@lucbv Yes, once we iron out all of bugs, we should definitely create a test so the ABL statistics can be tested in future releases.

**Comment by sayerhs** [comment]
@lawrenceccheung multiplication by `rho` in line 766 is the likely culprit https://github.com/Exawind/nalu-wind/blob/9b96b0af51b70459fe3e83ddf09160b7de3e5486/src/TurbulenceAveragingPostProcessing.C#L758-L770

**Comment by lawrenceccheung** [comment]
@sayerhs Thanks Shreyas.  
One question, though.  In [BdyLayerStatistics](https://github.com/Exawind/nalu-wind/blob/master/src/wind_energy/BdyLayerStatistics.C#L349), it's pulling the time average from the resolved averages:
```c++
  const auto velTimeAvg = nalu_ngp::get_ngp_field(meshInfo, "velocity_resa_abl");
```
Does that mean the resolved quantities in [line 800](https://github.com/Exawind/nalu-wind/blob/master/src/TurbulenceAveragingPostProcessing.C#L800) should also not have a rho?
```c++
    // Resolved quantities
      offset += numFavrePairs;
      for (int i=0; i < numResolvedPairs; ++i) {
        const int idx = offset + i;
        const auto prim = fieldPairs(idx).first.field;
        auto avg = fieldPairs(idx).second.field;
        const auto numComponents = fieldPairs(idx).first.scalarsDim1;

        for (unsigned j=0; j < numComponents; ++j) {
          const double avgVal = (
            avg.get(mi, j) * oldTimeFilter * zeroCurrent
            + rho * prim.get(mi, j) * dt) / currentTimeFilter;
          avg.get(mi, j) = avgVal;
        }
      }
```

**Comment by sayerhs** [comment]
@lawrenceccheung Pre-NGP all resolved quantities were computed with multiplication of density. So we can either divide by `density_ra` at this point in `TurbulentPostProcessing` as is done with Favre, or not skip the multiplication by `rho` and normalization by `rhoAvg` in 
https://github.com/Exawind/nalu-wind/blob/9b96b0af51b70459fe3e83ddf09160b7de3e5486/src/wind_energy/BdyLayerStatistics.C#L393-L396

https://github.com/Exawind/nalu-wind/blob/9b96b0af51b70459fe3e83ddf09160b7de3e5486/src/wind_energy/BdyLayerStatistics.C#L445-L448

**Comment by sayerhs** [comment]
@lawrenceccheung For reference here is the pre-NGP `BdyLayerStatistics.C` 

https://github.com/Exawind/nalu-wind/blob/db057d8f620b0b4da58029fe91075cb5b02b4c94/src/wind_energy/BdyLayerStatistics.C#L320-L366

**Comment by lawrenceccheung** [comment]
@sayerhs Yeah, I can just change the calculation in `BdyLayerStatistics.C` to keep things consistent with pre-NGP and try to minimize the confusion.  

But there's always a finite probability that any change I make will break something else unintentionally, so we should test out these statistics comprehensively...

**Comment by lawrenceccheung** [comment]
I tested the changes and they seem to address the issues I was seeing.   The changes are in the pull request [506](https://github.com/Exawind/nalu-wind/pull/506), but @sayerhs you may have gotten to it already.

Lawrence

**Comment by sayerhs** [comment]
@lawrenceccheung I tried to address that and another issue in RANS averaging in this commit 4c430563cd3b2c9443d871874b174132cd51fe4c
